`#[account]` does not work in doctest

* **premise:** i want to define an `#[account] struct ...` in a doctest.

* **setup:**

```rust
//! Module docstring fails:
//! ```
//! # use anchor_lang::prelude::*;                                                                                                                  
//! # declare_id! { "TESTESTESTESTESTESTESTESTESTESTTTTTTTTTTEST" }                                                                                 
//! #[account] struct Fail1 {}
//! ```

/// Item docstring fails:
/// ```
/// # use anchor_lang::prelude::*;                                                                                                                  
/// # declare_id! { "TESTESTESTESTESTESTESTESTESTESTTTTTTTTTTEST" }                                                                                 
/// #[account] struct Fail2 {}    
/// ```
struct Stub;
```

```
cargo test
```

* **expected:** compiles (and does nothing.)

* **actual:**

```
error[E0425]: cannot find value `ID` in the crate root                                                                                                  
  = note: this error originates in the attribute macro `account` (in Nightly builds, run with -Z macro-backtrace 
Couldn't compile the test.
```

**proximate cause:** https://github.com/solana-foundation/anchor/blob/5300d7cf8aaf52da08ce331db3fc8182cd821228/lang/syn/src/codegen/program/idl.rs#L32 `cargo test` compiles doctests as separate crate. the code of each doctest does not live at the top level of the doctest crate. that makes it impossible to provide `crate::ID` to the `#[account]` macro.

--- a/lang/attribute/account/src/lib.rs
+++ b/lang/attribute/account/src/lib.rs
@@ -171,7 +171,11 @@ pub fn account(
                 #[automatically_derived]
                 impl #impl_gen anchor_lang::Owner for #account_name #type_gen #where_clause {
                     fn owner() -> Pubkey {
-                        crate::ID
+                        // In a doctest the ID will be in the current scope, not the crate root
+                        #[cfg(not(doctest))]
+                        { crate::ID }
+                        #[cfg(doctest)]
+                        { ID }
                     }
                 }
             }

--- a/lang/attribute/event/src/lib.rs
+++ b/lang/attribute/event/src/lib.rs
@@ -177,7 +177,11 @@ pub fn emit_cpi(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
                 .collect();
 
             let ix = anchor_lang::solana_program::instruction::Instruction::new_with_bytes(
-                crate::ID,
+                // In a doctest the ID will be in the current scope, not the crate root
+                #[cfg(not(doctest))]
+                { crate::ID },
+                #[cfg(doctest)]
+                { ID },
                 &ix_data,
                 vec![
                     anchor_lang::solana_program::instruction::AccountMeta::new_readonly(

--- a/lang/syn/src/codegen/accounts/mod.rs
+++ b/lang/syn/src/codegen/accounts/mod.rs
@@ -26,8 +26,18 @@ pub fn generate(accs: &AccountsStruct) -> proc_macro2::TokenStream {
     let impl_dup_mutable_keys = duplicate_mutable_account_keys::generate(accs);
     let bumps_struct = bumps::generate(accs);
 
-    let __client_accounts_mod = __client_accounts::generate(accs, quote!(crate::ID));
-    let __cpi_client_accounts_mod = __cpi_client_accounts::generate(accs, quote!(crate::ID));
+    let program_id = quote! {
+        // In a doctest the ID will be in the current scope, not the crate root
+        {
+            #[cfg(not(doctest))]
+            { crate::ID }
+            #[cfg(doctest)]
+            { ID }
+        }
+    };
+
+    let __client_accounts_mod = __client_accounts::generate(accs, program_id.clone());
+    let __cpi_client_accounts_mod = __cpi_client_accounts::generate(accs, program_id);
 
     let ret = quote! {
         #impl_try_accounts

